### PR TITLE
Remove Buffer Interleave

### PR DESCRIPTION
**Pull request by srsampson**
The interleave has not been used by anyone, and complicates the code.

### CODE REVIEW

**Comment by srsampson** [comment]
Note - It will be awhile before I expect Travis to start passing.

**Comment by drowe67** [comment]
@srsampson - think we need to plan this one carefully:

1. We want to keep the gold primer interleaver - that works fine and is necessary.  We can use that to scramble symbols within codewords, and provide protection from fading
1. The code I want to remove is around interleaving over multiple codewords, and the interleaver state machine.
1. Try to make very small changes, and make sure the ctests pass at every step.  Test locally first, before pushing.  Large changes will lead to unfixable problems we have to back out of.
1. For this PR - suggest you start with changes to **just** ofdm_mod.c - e.g. remove the --interleave cmd line option, then make sure all ctests pass.  May have to tweak tests or remove any ctests that use the --interleave option.  Most of our work uses interleave 1 - so should still work, and will give you something to test against.
1. It's possible to make these changes without committing any code that break the tests - and will result in much faster development.  The tests can help you.

I have to got to my day job now, but can help you plan this one carefully later.

**Comment by srsampson** [comment]
Oh OK, I misunderstood. I was thinking we were backing the whole thing out.

**Comment by srsampson** [comment]
I'm doing a little more than you wanted, but trying to kill two birds...

Finished for the night, and will do the API's in the morning.

I'm calling it "buffered interleave" removal, as opposed to "scrambled interleave" that we want to keep.

**Comment by srsampson** [comment]
All tests are passing, but since I got rid of the unused "advanced" structure in the open call, the fdmdv GUI is failing.

**Comment by srsampson** [comment]
Errors outstanding:
```
/home/travis/freedv-gui/src/fdmdv2_main.cpp: In member function ‘virtual void MainFrame::OnTogBtnOnOff(wxCommandEvent&)’:
/home/travis/freedv-gui/src/fdmdv2_main.cpp:2740:40: error: aggregate ‘MainFrame::OnTogBtnOnOff(wxCommandEvent&)::freedv_advanced adv’ has incomplete type and cannot be defined
                 struct freedv_advanced adv;
                                        ^
/home/travis/freedv-gui/src/fdmdv2_main.cpp:2742:62: error: ‘freedv_open_advanced’ was not declared in this scope
                 g_pfreedv = freedv_open_advanced(g_mode, &adv);
                                                              ^
/home/travis/freedv-gui/src/fdmdv2_main.cpp: In function ‘void per_frame_rx_processing(FIFO*, FIFO*)’:
/home/travis/freedv-gui/src/fdmdv2_main.cpp:4266:75: error: ‘freedv_get_sync_interleaver’ was not declared in this scope
             g_interleaverSyncState = freedv_get_sync_interleaver(g_pfreedv);
                                                                           ^
[ 29%] Building CXX object src/CMakeFiles/freedv.dir/fdmdv2_pa_wrapper.cpp.o
src/CMakeFiles/freedv.dir/build.make:127: recipe for target 'src/CMakeFiles/freedv.dir/fdmdv2_main.cpp.o' failed
make[2]: *** [src/CMakeFiles/freedv.dir/fdmdv2_main.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:90: recipe for target 'src/CMakeFiles/freedv.dir/all' failed
make[1]: *** [src/CMakeFiles/freedv.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by srsampson** [comment]
FYI - This pull updates the freedv config structure to include the new variables to be used with packet and QAM.

File: src/freedv_api.c [hunk 1]
```diff
@@ -103,11 +103,13 @@ char *rx_sync_flags_to_text[] = {
...
-struct freedv *freedv_open(int mode) {
-    return freedv_open_advanced(mode, NULL);
-}
+// Obsolete call
 
 struct freedv *freedv_open_advanced(int mode, struct freedv_advanced *adv) {
+    return freedv_open(mode);
+}
+
+struct freedv *freedv_open(int mode) {
```
Comment (drowe67) [inline src/freedv_api.c:112]: Nice @srsampson - you haven't changes the API, despite having an obsolete call!  :+1:  Removing _advanced would break many end user applications....

**Comment by drowe67** [comment]
Wow - that's a lot of work -and all tests passing  well done @srsampson :1st_place_medal: 

There's a lot of changes here so I'll need a few days to review :smile: 

I guess you saw how messy that interleaving code made the C code.  It was something that showed some promise in the Octave simulations but wasn't effective in real time.  Good experience I guess.  The modem frame/packet model and longer LDPC codewords is working better for the data modes.

**Comment by srsampson** [comment]
I'm glad you reminded me that the interleave is both a buffer concept and a burst error concept (scrambling the symbols around).  I'm glad my other PR blew up! ha.

Yes, I tried to be careful when pulling the buffering stuff. I know the test passing is about 50% of the job! ha...

**Comment by drowe67** [comment]
Re merging this one - well as the API hasn't changed freedv-gui is linking OK (Travis tried that for us).  freedv-gui does have an interleaver field, I suspect that will be doing nothing atm.  For completeness we should probably remove that field from freedv-gui.  For other peoples applications that use the FreeDV API - a benign fail (like ignoring the interleaver value but not breaking) is a good thing.

**Comment by srsampson** [comment]
I say merge it, and let's see how long it takes for someone to say "hey, this buffer interleaver is a gimick!" ha

**Comment by srsampson** [comment]
Selecting mode 2020 on FreeDV causes the GUI to crash with:
```
Start .....
freedv: /home/ssampson/codec2/src/ofdm.c:1576: ofdm_set_sync: Assertion `ofdm != NULL' failed.
Aborted (core dumped)
```

**Comment by srsampson** [comment]
Re-examined my setup here and it was kind of dirty. Turns out I had a stray codec2 library file in /usr/lib

Building freedv with this branch does produce output. I think I need a second opinion though, as I thought the audio was not as good as the master codec2. Seemed to be more missed decodes by my ear. Haven't compared the numbers yet.

Will do more testing today to better compare.

**Comment by srsampson** [comment]
Comparing master to ss-interleave the "stats" display is identical on 700D. No errors except at beginning and end as the data starts up and closes down.

I think ss-interleave ready for a merge. I'll close ss-modem and will start a new branch to just work the packet length option for data mode.

**Comment by drowe67** [comment]
> Comparing master to ss-interleave the "stats" display is identical on 700D. No errors except at beginning and end as the data starts up and closes down.
> 
> I think ss-interleave ready for a merge. I'll close ss-modem and will start a new branch to just work the packet length option for data mode.

The idea of merging ss-modem first is to contain the changes in small chunks.  Mega PRs with many files are hard to review, and harder to back out of/debug if anything goes wrong.  Mega-PRs also tend to get out of sync with master and other code being merged into master which leads to conflicts.  All things I've done :roll_eyes: 

OK so - 700D and 2020 worked OK on freedv-gui even with the interleaver disabled?  Cool :+1: 

I'll review the code over the next few days, then we can merge :slightly_smiling_face:

**Comment by drowe67** [comment]
@srsampson I'll fix those conflicts for you - back in 10 !

**Comment by drowe67** [comment]
Hi @srsampson while I was on a roll I also tested this PR https://github.com/drowe67/codec2/pull/138/commits/cdd75ffe4ab93d7ae67b893c1e094be5a403820a with freedv-gui.  Same results - as you - it works on 700D and 2020 and ignores the Interleaver setting :grinning: :1st_place_medal: That's good news for anyone else using the FreeDV API.

File: src/freedv_api.c [hunk 2]
```diff
@@ -649,7 +613,7 @@ int freedv_comprx(struct freedv *f, short speech_out[], COMP demod_in[]) {
     }
 
     if (FDV_MODE_ACTIVE( FREEDV_MODE_700D, f->mode)) {
-        rx_status = freedv_comp_short_rx_700d(f, (void*)demod_in, 0, 1.0);
+        rx_status = freedv_comp_short_rx_700d(f, (void*)demod_in, 0, 2.0f); // was 1.0 ??
```
Comment (drowe67) [inline src/freedv_api.c:616]: Why did you change this?

File: src/interldpc.c [hunk 3]
```diff
@@ -163,43 +163,12 @@ void qpsk_modulate_frame(COMP tx_symbols[], int codeword[], int n) {
...
-            /* sucessful decode! */
-            next_sync_state_interleaver = synced;
-            ofdm->frame_count_interleaver = interleave_frames;
-        }
-    }
-
-    ofdm->sync_state_interleaver = next_sync_state_interleaver;
-}
-
 /* measure uncoded (raw) bit errors over interleaver frame, note we
```
Comment (drowe67) [inline src/interldpc.c:166]: Oh boy - nice to see this one go :+1:   Was so much trouble to get going, added a lot of complexity.  Oh well, that's experimental radio!

File: src/freedv_api.c (hunk 2 above)
Comment (srsampson) [reply src/freedv_api.c:616]: I noticed everywhere else it was set to 2.0 (about 2 other places if I recall), so I thought it should be the same.

Comment (drowe67) [reply src/freedv_api.c:616]: Guess it's just a gain factor, which doesn't matter much for QPSK ...

File: src/interldpc.c (hunk 3 above)
Comment (srsampson) [reply src/interldpc.c:166]: In freedv_api,c the call just returns true:
```
int freedv_get_sync_interleaver(struct freedv *f) {
    return 1;
}
```

**Comment by drowe67** [comment]
Nice work @srsampson - so much code removed!  

OK given we with messed with freedv_api (and also recently the ofdm modem), I wanted to run the stm32 ctests.  Have modified a few tests that weren't building, but the stm32 ctests are failing.  They were all OK about a month ago (v4 SM1000 code) so it will be in the latest changes.  I'll have another look at it in my AM.

**Comment by srsampson** [comment]
I made some changes in accordance with the interleave, but some tests still failing.
```
The following tests FAILED:
	  9 - tst_ofdm_demod_AWGN (Failed)
	 10 - tst_ofdm_demod_fade (Failed)
	 12 - tst_ofdm_demod_ldpc_AWGN (Failed)
	 13 - tst_ofdm_demod_ldpc_fade (Failed)
	 14 - tst_codec2_enc_1300 (Failed)
	 15 - tst_codec2_enc_700C (Failed)
	 16 - tst_codec2_dec_1300 (Failed)
	 17 - tst_codec2_dec_700C (Failed)
	 18 - tst_api_mod_700D_TEST (Failed)
	 19 - tst_api_mod_700D_CODEC (Failed)
``

**Comment by drowe67** [comment]
Wow that's great progress @srsampson - well done :+1:   I'll take a look at where you are up to

**Comment by srsampson** [comment]
I'm almost thinking that throwing out that ofdm_demod in unittest and just
using the one in src as it looks a lot cleaner.

I think it's failing because it's getting not-a-number calculation on BER.

**Comment by drowe67** [comment]
@srsampson - I've fixed it - one minute :-)

**Comment by drowe67** [comment]
BTW - you need to sleep more - you're going to bed after I am (in Australian time) :grinning:

**Comment by drowe67** [comment]
I think we're down to:
```
tst_ofdm_demod_fade
tst_ofdm_demod_ldpc_fade
```
BTW:
```
ctest -R tst_codec2_
` ``
all passed for me (you mentioned them failing for you).  Maybe try them again at your end.

**Comment by drowe67** [comment]
The logs are quite useful (thanks @donmr :slightly_smiling_face: ):
```
david@bear:~/codec2/stm32/unittest/test_run/tst_ofdm_demod_fade$ cat setup.log
cohpsk_ch ----------------------------------------------------------------------------------
Fs: 8000 NodB: -24.00 foff: -10.00 Hz fading: 2 inclip: 1.00 ssbfilt: 1 complexout: 0
cohpsk_ch ----------------------------------------------------------------------------------
-----------------------------------------------------
cohpsk_ch ERROR: Can't find fading file: /home/david/codec2/raw/fast_fading_samples.float
->See cohpsk_ch.c source for instructions on how to generate this file.
-----------------------------------------------------
```

**Comment by srsampson** [comment]
I seem to be back on Saudi Arabia time. I was doing my laundry and making
my bed at 3am.

I slept until it got too hot and I can't seem to take a lot of air
conditioning without getting a sore throat.

I'm waiting for 7pm when I'll go mow the lawn. It should be below 80 deg by
then.

**Comment by drowe67** [comment]
OK most of them passing for me now, just these two left:
```
18/23 Test #18: tst_api_mod_700D_TEST ............***Failed   42.90 sec
      Start 19: tst_api_mod_700D_CODEC
19/23 Test #19: tst_api_mod_700D_CODEC ...........***Failed   42.84 sec
```
Curiously, their `_demod` counterparts pass.

I backed off the noise on the fading tests to get a pass.  Possible explanations:
1. We broke the modem and it's performing worse.  However I'm pretty sure this would have been picked up by the x86 tests.  Or will be soon as other development continues.
1. The modem changes (like interleaver removal) means it's sync behaviour has changed a bit, so it's locking up early and copping a low SNR section of the simulated channel.
1. We just happen to be getting a couple of dud frames with high BERs.  The stm32 tests have to be short as they are so slow to run, so it's easy to hit a bad patch of channel that skews the results.
1. The important thing is the stm32 and x86 modems have identical results:
   ```
   david@bear:~/codec2/stm32/unittest/test_run$ cat tst_ofdm_demod_fade/check.log
   /home/david/codec2/stm32/unittest/test_run/tst_ofdm_demod_fade

   BER checks
   Reference
   BER......: 0.0941 Tbits: 85680 Terrs:  8062
   BER2.....: 0.0903 Tbits: 80920 Terrs:  7304
   Target
   BER......: 0.0941 Tbits: 85680 Terrs:  8062
   BER2.....: 0.0903 Tbits: 80920 Terrs:  7304
   PASS: max BER
   PASS: BER compare

   Test PASSED
   ```
   The main purpose of the stm32 tests is to check perf is close to x86 - not debug the modem.  If there are any modem issues I'm sure we'll catch them in x86 land at some point.

**Comment by srsampson** [comment]
In:
scripts/run_stm32_prog

change:
        monitor adapter_khz 4000

to:

     monitor adapter speed 4000

The former is deprecated.

**Comment by srsampson** [comment]
> BTW:
> 
> ```
> ctest -R tst_codec2_
> ```
> 
> all passed for me (you mentioned them failing for you). Maybe try them again at your end.

I had to sudo cp the libraries to /usr/local/lib

and the codec tests started passing

FYI

**Comment by drowe67** [comment]
Ok so looking at the difference between in the two mod output files in `.../test_run/tst_api_mod_700D_TEST`, which are ```ref_mod.raw``` and ```stm_out.raw```, I can see some differences at frame rate:

Screenshot from 2020-07-24 07-58-14

Cyclic prefix thing?

I might take a break now, and all my commits are up to date, so feel free to work on it (if you want to) :slightly_smiling_face:

**Comment by drowe67** [comment]
> I had to sudo cp the libraries to /usr/local/lib

Hmm, we don't want undocumented traps like that laying around ... I seem to be able to run them with no local installation:
```
ldconfig -v 2>/dev/null | grep codec2
(silence)
ctest -R tst_codec2_
Test project /home/david/codec2/stm32/build_stm32
    Start 14: tst_codec2_enc_1300
1/4 Test #14: tst_codec2_enc_1300 ..............   Passed   58.84 sec
    Start 15: tst_codec2_enc_700C
2/4 Test #15: tst_codec2_enc_700C ..............   Passed   58.93 sec
    Start 16: tst_codec2_dec_1300
3/4 Test #16: tst_codec2_dec_1300 ..............   Passed   26.93 sec
    Start 17: tst_codec2_dec_700C
4/4 Test #17: tst_codec2_dec_700C ..............   Passed   25.92 sec
```

**Comment by donmr** [comment]
On some of those tests I have to choose how close was a "pass" based
on experiments.  The settings may need to be adjusted.

Don

**Comment by srsampson** [comment]
The error I'm getting on the 700D_TEST shows the raw file comparison is way off. Not even close.

**Comment by drowe67** [comment]
> The error message I got in the log was could not find the library libcodec2.so.9 or some such.
> 
> I was surprised as well, as it should have used the one in the build, not one for who knows what.

Could be a path order thing - it might be finding codec2 utilities installed on your system first.  We'll keep an eye on it :detective:

**Comment by drowe67** [comment]
> The error I'm getting on the 700D_TEST shows the raw file comparison is way off. Not even close.

That's what I saw initially - then I plotted them and found out it was just chunks of samples at the frame rate.  If its a CP thing it could be related to the change in fading channel tests.

**Comment by drowe67** [comment]
Hi @srsampson - I've dug a little more into the `ctest -V -R tst_api_mod_700D_TEST` that is failing:
1. It's the last 160 samples of every frame.
1. `check.log` indicates that despite being different, both files demodulate to the same bits, error free:
   ```
   Reference check
   frames decoded: 24  output speech samples: 2560
   BER......: 0.0000 Tbits:  5456 Terrs:     0
   Coded BER: 0.0000 Tbits:  2464 Terrs:     0
   Passed

   Target check
   frames decoded: 24  output speech samples: 2560
   BER......: 0.0000 Tbits:  5456 Terrs:     0
   Coded BER: 0.0000 Tbits:  2464 Terrs:     0
   Passed

   Compare output binary data
   1120 -4590 -5676
   1121 -4588 -910
   1122 792 2359
   <snip>
   ```
1. I've started trying earlier versions to bracket the problem - https://github.com/drowe67/codec2/pull/111/commits/210a451a7475b9f0c6490e5c00847370a4a69a0f from https://github.com/drowe67/codec2/pull/111 passes.  You need to be sure to rebuild the host side as well as stm32.

**Comment by drowe67** [comment]
@srsampson OK It looks the problem starts at ab3a22ad68 in https://github.com/drowe67/codec2/pull/114 (`dr-ofdm-rm-statics` branch).  I had to push a small change ab3a22ad to get a couple of other tests to build. 

Can you pls tell me if you can repeat these results?

**Comment by srsampson** [comment]
It will be a couple of hours as I'm about to run the mower. Damned chores
interfering with fun!

**Comment by srsampson** [comment]
I ran the 700D_TEST and 700D_CODEC on the branch you mentioned ab3a22a and yes, the tests fail there as well.

I couldn't run all of the tests because "tst_ofdm_demod_quick" seemed to be in an endless loop.

**Comment by drowe67** [comment]
> I couldn't run all of the tests because "tst_ofdm_demod_quick" seemed to be in an endless loop.

Yep many are broken for reasons that we fix already in this PR, I've just been focussing on `ctest -V -R tst_api_mod_700D_TEST`.

I saved the `ref_mod.raw` and `stm_out.raw` files from both branches to compare.  I'm seeing the x86 `ref_mod.raw` change between branches - the stm32 version stays the same. So hopefully we can debug this one in x86 land which should make it easier.  the script that generates the reference starts [here](https://github.com/drowe67/codec2/blob/master/stm32/unittest/scripts/tst_api_mod_setup#L28).  However it might produce the same problem using any speech source.

**Comment by drowe67** [comment]
Ok I can reproduce it on a x86 with `make && ./src/freedv_tx 700D ../raw/hts1a.raw hts1a_mod_head.raw`
It's not in https://github.com/drowe67/codec2/pull/114, but some code that was merged into it towards the end via https://github.com/drowe67/codec2/pull/101 `horusbinary_v2`
Have tracked it down to between  `8721cf6` and `970af78`, not sure what branch/PR that is yet

**Comment by drowe67** [comment]
Doh! A few PRs ago I simplified [freedv_tx.c](https://github.com/drowe67/codec2/blob/ss-interleave/src/freedv_tx.c), removing the text transmission, which is in the last few symbols of the frame.  the stm32 modulator test still has the text transmission, so the waveforms didn't match, but the BER decode worked perfectly, as we don't check text characters.

That's a day of my life I'll never get back :roll_eyes: but learnt a bit more about wild Git merge paths ....

So wasn't ever anything to do with this PR :+1: although it was good to clean up the stm32 tests, as quite a bit has changed recently.

**Comment by drowe67** [comment]
Pls let me know if there is anything else you would like to do in this PR, otherwise I'll merge :+1:

**Comment by srsampson** [comment]
Great find! I'm a bit disappointed in how ss-modem and ss-interleave went, as it was supposed to be a simple side-affair to the real work of converting the new Octave code. Then the git snafu kind of put a fork in the road. I need to remember to check on stm32, as I completely skipped that check. All water under the bridge now:
```
codec2/build_linux
100% tests passed, 0 tests failed out of 59

Total Test time (real) =  96.41 sec

codec2/stm32/build_stm32
100% tests passed, 0 tests failed out of 23

Total Test time (real) = 808.33 sec
```
There's 904.74 seconds I'll never get back! ha. Ready to merge on my end.
Thanks!!

**Comment by srsampson** [comment]
Just a quick check, but I burned the image to the SM1000 and it boots, function switches operate, and PTT works without ring of death.

**Comment by drowe67** [comment]
@srsampson Actually I was impressed with how fast you developed this PR - it was a big job, so well done :+1: 

The stm32 tests don't run automatically with every push of the x86 code, which means they can get behind if we let them, oh well.  Yeah I get myself it Git tangles all the time, but I was petty grateful for source control when bracketing that bug above.

In https://github.com/drowe67/codec2/pull/126#issuecomment-656939078 we planned some tasks for the modem.  You've actually done the big one!  Would you like to try either (1) or (2)?  I'm happy to take one of them too if you like.  No rush BTQ, just if and when you feel like it :slightly_smiling_face: 

In https://github.com/drowe67/codec2/pull/136, I've added a bunch of new LDPC codes from @billcowley - so we have these available for any new data modes too.  They're much easier to use now via the functions in https://github.com/drowe67/codec2/blob/master/src/ldpc_codes.c, see https://github.com/drowe67/codec2/blob/master/src/ldpc_enc.c for an example.
